asynciterator.ts

```diff
@@ -1852,17 +1894,166 @@ class HistoryReader<T> {
+
+  /**
+   *
+   * @param source
+   */
+  constructor(source: Iterator<T>) {
+    super();
+    this._source = source;
+    this.readable = true;
+  }
+
+  read(): T | null {
+    let value: T | null = null;
+    if (this._source) {
+      const item = this._source.next();
+      if (!item.done)
+        value = item.value;
+      else
+        this.close();
+    }
+    return value;
+  }
+
+  protected _end(destroy: boolean = false) {
+    super._end(destroy);
+    delete this._source;
+  }
+}
+
+/**
+ * The WrappingIterator class can be used to wrap all supported kinds of data
+ * sources, both promisified and not, into AsyncIterator instances.
+ * This class is not meant to be used directly but, rather, through the
+ * `wrap()` function as it provides the mechanism that allows the latter
+ * to synchronously return an instance of AsyncIterator even for promisified
+ * sources.
+ */
+export class WrappingIterator<T> extends AsyncIterator<T> {
+  protected _source?: AsyncIteratorLike<T>;
+
+  constructor(sourceOrPromise: WrapSource<T> | Promise<WrapSource<T>> | null | undefined, options: WrapOptions = {}) {
+    super();
+    this._onSourceEnd = this._onSourceEnd.bind(this);
+    this._onSourceError = this._onSourceError.bind(this);
+    this._onSourceReadable = this._onSourceReadable.bind(this);
+    if (sourceOrPromise)
+      this.setSource(sourceOrPromise, options);
+  }
+
+  setSource(sourceOrPromise: WrapSource<T> | Promise<WrapSource<T>>, options: WrapOptions = {}) {
+    if (isPromise(sourceOrPromise)) {
+      sourceOrPromise
+        .then(source => this._wrapSource(source, options))
+        .catch(err => this.emit('error', err));
+    }
+    else {
+      this._wrapSource(sourceOrPromise, options);
+    }
+  }
+
+  protected _wrapSource(source: WrapSource<T>, options: WrapOptions) {
+    try {
+      const wrappedSource = isAsyncIteratorLike<T>(source) ? source : fromSource<T>(source, options);
+      wrappedSource.on('end', this._onSourceEnd);
+      wrappedSource.on('error', this._onSourceError);
+      wrappedSource.on('readable', this._onSourceReadable);
+      this._source = wrappedSource;
+      this.readable = true;
+    }
+    catch (err) {
+      scheduleTask(() => this.emit('error', err));
+    }
+  }
+
+  protected _onSourceReadable() {
+    if (this.readable)
+      this.emit('readable');
+    else
+      this.readable = true;
+  }
+
+  protected _onSourceEnd() {
+    this.close();
+  }
+
+  protected _onSourceError(err: Error) {
+    this.emit('error', err);
+  }
+
+  protected _end(destroy: boolean = false) {
+    super._end(destroy);
+    if (this._source) {
+      this._source.removeListener('end', this._onSourceEnd);
+      this._source.removeListener('error', this._onSourceError);
+      this._source.removeListener('readable', this._onSourceReadable);
+      delete this._source;
+    }
+  }
+
+  read(): T | null {
+    if (this._source)
+      return this._source.read();
+    return null;
```

Comment:
Why not use a ternary here?

Reply:
I find ternary iterators to be somewhat difficult to handle on the debugging side of things (as in when using a debugger). My first instinct is to use them, which is why you'll find them in other code bits of mine, but then I often come to regret it when debugging and thus I'm trying to force myself not to use them anymore.